s3fs/tests/test_s3fs.py

```diff
@@ -466,6 +466,21 @@ def test_new_bucket(s3):
     with pytest.raises((IOError, OSError)):
         s3.ls('new')
 
+def test_dynamic_add_rm(s3):
+    s3.mkdir('one')
+    s3.mkdir('one/two')
+    assert s3.exists('one')
+    # The following ls is currently causing s3fs to tag 'one/two' as an empty
+    # key but NOT as a directory. 
+    s3.ls('one')
+    s3.touch("one/two/file_a")
+    assert s3.exists('one/two/file_a')
+    # Now 'one/two' should be tagged as a directory since it contains a file
+    # but this recursive rm does NOT find file_a unless another 'ls' is performed:
+    #s3.ls('one', refresh=True) #Retags 'one/two' as a DIRECTORY.
```

Comment:
Okay, I'll try to try that out sometime this week. Might be able to modify the mkdir function to update the cache correctly so that any newly created directories are flagged as such in the cache.